xmtp_cursor_state/src/store.rs

```diff
@@ -0,0 +1,227 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::{Arc, Mutex};
+use xmtp_proto::xmtp::xmtpv4::envelopes::Cursor;
+
+pub type SharedCursorStore = Arc<Mutex<CursorStore>>;
+
+pub struct CursorStore {
+    topics: HashMap<Vec<u8>, Cursor>,
+}
+
+impl CursorStore {
+    pub fn new() -> Self {
+        Self {
+            topics: HashMap::new(),
+        }
+    }
+
+    /// Record that a message for this topic with the given clock was processed
+    pub fn processed(&mut self, topic: Vec<u8>, new_clock: &Cursor) {
+        let current = self.topics.entry(topic).or_default();
+        current.merge_cursor(new_clock);
+    }
+
+    /// Get the current vector clock for this topic
+    pub fn get_latest(&self, topic: &[u8]) -> Option<&Cursor> {
+        self.topics.get(topic)
+    }
+
+    /// Compute the lowest common cursor across a set of topics.
+    /// For each node_id, uses the **minimum** sequence ID seen across all topics.
+    pub fn lowest_common_cursor(&self, topics: &[Vec<u8>]) -> Option<Cursor> {
```

Comment:
this one is worth discussing a bit more. It is obviously not efficient to re-load a bunch of messages that we have seen already, just because a new topic was added. Handling this correctly is a large project. Doing the LCC is guaranteed to be correct. We need to discuss what happens if the client receives duplicates. Fundamentally last_seen when used with many groups has a "at-least-once" semantic. Not sure what that means for the rest of the system.

Reply:
If we receive a duplicate, won't we know its a duplicate based on what's in persistent storage?

Reply:
We did discuss this last year, what do you think of this? https://www.notion.so/ephemerahq/Efficient-and-reliable-batch-subscriptions-10a30823ce928060b6a0cf4e38ee7138?source=copy_link

Reply:
Ah this is my bad, I totally misunderstood what was being discussed here! You brought up a great point about deduping the 'logical' payload. I wonder if it would make sense to dedupe against [this field](https://github.com/xmtp/proto/blob/637b892e32c5409a57f18c84291fc9007b4fb160/proto/xmtpv4/envelopes/envelopes.proto#L41), or a hash of it? That's assuming that the `depends_on` in the AAD won't be different between re-publishes. We could maybe structure it as a unique index in the sqlite database?

Reply:
the entire `ClientEnvelope` will be the "same". The wrapping PayerEnvelope and OriginatorEnvelope will be different.

Not sure whether it is guaranteed that the serialized bytes will be exactly the same. I would not want to depend on it. The client might decide to republish something via a different payer.

Reply:
I like that!